# Layoffs Analysis 

The project aim was to derive insights from a layoffs dataset using SQL for extensive **Exploratoty Data Analysis (EDA)**.

In [1]:
%load_ext sql

In [2]:
%sql mysql+pymysql://root: **** @localhost:3306/worldlayoffs
# Connecting to the database

'Connected: root@worldlayoffs'

In [3]:
%%sql
SHOW TABLES;

 * mysql+pymysql://root:***@localhost:3306/worldlayoffs
3 rows affected.


[('layoffs',), ('layoffs_staging',), ('layoffs_staging2',)]

This part will use the `layoffs_staging2` table cleaned in the previous queries. (from the attached `layoffs_Data_Cleaning` Notebook)<br> 
Having an overview of the data in the `layoffs_staging2` table.

In [4]:
%%sql
SELECT * 
FROM layoffs_staging2
LIMIT 5;

 * mysql+pymysql://root:***@localhost:3306/worldlayoffs
5 rows affected.


[('Included Health', 'SF Bay Area', 'Healthcare', None, '0.06', datetime.date(2022, 7, 25), 'Series E', 'United States', 272),
 ('&Open', 'Dublin', 'Marketing', 9, '0.09', datetime.date(2022, 11, 17), 'Series A', 'Ireland', 35),
 ('#Paid', 'Toronto', 'Marketing', 19, '0.17', datetime.date(2023, 1, 27), 'Series B', 'Canada', 21),
 ('100 Thieves', 'Los Angeles', 'Consumer', 12, None, datetime.date(2022, 7, 13), 'Series C', 'United States', 120),
 ('10X Genomics', 'SF Bay Area', 'Healthcare', 100, '0.08', datetime.date(2022, 8, 4), 'Post-IPO', 'United States', 242)]

## Exploratory Data Analysis (EDA) 

### Objectives
1. Top 10 companies with maximum layoffs
2. Top 5 companies that experienced total shutdown
3. Top 10 countries where most layoffs happened
4. Top 10 locations in USA with most layoffs
5. Company stage with the most layoffs
6. The industry with the most layoffs
7. Total layoffs per year
8. Top 5 companies with maximum layoffs year-wise
9. Monthly layoffs rolling tota

### 1. Top 10 companies with maximum layoffs

In [5]:
%%sql
SELECT company, 
    SUM(total_laid_off) as sum_laid_off_per_company
FROM layoffs_staging2
GROUP BY company
ORDER BY 2 DESC
LIMIT 10;

 * mysql+pymysql://root:***@localhost:3306/worldlayoffs
10 rows affected.


[('Amazon', Decimal('18150')),
 ('Google', Decimal('12000')),
 ('Meta', Decimal('11000')),
 ('Salesforce', Decimal('10090')),
 ('Microsoft', Decimal('10000')),
 ('Philips', Decimal('10000')),
 ('Ericsson', Decimal('8500')),
 ('Uber', Decimal('7585')),
 ('Dell', Decimal('6650')),
 ('Booking.com', Decimal('4601'))]

In [6]:
%%sql
SELECT MIN(`date`) as minimum_date,
    MAX(`date`) as maximum_date
FROM layoffs_staging2;

 * mysql+pymysql://root:***@localhost:3306/worldlayoffs
1 rows affected.


[(datetime.date(2020, 3, 11), datetime.date(2023, 3, 6))]

The output above shows that the sum of layoffs in the given companies happened between March 2020 and March 2023. With Amazon company experiencing the maximum layoffs with a total of 36300 layoffs in 3 years. The top 10 companies that experienced maximum layoffs are Amazon, Google, Meta, Salesforce, Microsoft, Philips, Ericsson, Uber, Dell, and Booking.com.

### 2. Top 5 companies that experienced total shutdown

In [7]:
%%sql 
SELECT * 
FROM layoffs_staging2
WHERE percentage_laid_off = 1
ORDER BY total_laid_off DESC
LIMIT 5;

 * mysql+pymysql://root:***@localhost:3306/worldlayoffs
5 rows affected.


[('Katerra', 'SF Bay Area', 'Construction', 2434, '1', datetime.date(2021, 6, 1), 'Unknown', 'United States', 1600),
 ('Butler Hospitality', 'New York City', 'Food', 1000, '1', datetime.date(2022, 7, 8), 'Series B', 'United States', 50),
 ('Deliv', 'SF Bay Area', 'Retail', 669, '1', datetime.date(2020, 5, 13), 'Series C', 'United States', 80),
 ('Jump', 'New York City', 'Transportation', 500, '1', datetime.date(2020, 5, 7), 'Acquired', 'United States', 11),
 ('SEND', 'Sydney', 'Food', 300, '1', datetime.date(2022, 5, 4), 'Seed', 'Australia', 3)]

In [8]:
%%sql 
SELECT MAX(total_laid_off) as maximum_layoff,
    MAX(percentage_laid_off) as maximum_percentage_layoff
FROM layoffs_staging2;

 * mysql+pymysql://root:***@localhost:3306/worldlayoffs
1 rows affected.


[(12000, '1')]

From the above output, 12000  was the maximum number of people laid off in a single day. While having a value of 1 for the `percentage_laid_off` shows that the company fired everyone and probably got shut down. <br>
The top 5 companies that encountered total shutdown were Katerra, Butler Hospitality, Deliv, Jump, and SEND. Most of these companies were located in the United States

### 3. Top 10 countries where most layoffs happened

In [9]:
%%sql
SELECT country, 
    SUM(total_laid_off) as sum_laid_off_per_country
FROM layoffs_staging2
GROUP BY country
ORDER BY 2 DESC
LIMIT 10;

 * mysql+pymysql://root:***@localhost:3306/worldlayoffs
10 rows affected.


[('United States', Decimal('256559')),
 ('India', Decimal('35993')),
 ('Netherlands', Decimal('17220')),
 ('Sweden', Decimal('11264')),
 ('Brazil', Decimal('10391')),
 ('Germany', Decimal('8701')),
 ('United Kingdom', Decimal('6398')),
 ('Canada', Decimal('6319')),
 ('Singapore', Decimal('5995')),
 ('China', Decimal('5905'))]

The United States experienced the most layoffs with 256,559 employees being laid off. 

### 4. Top 10 locations in the USA with most layoffs

In [10]:
%%sql
SELECT 
    location,
    SUM(total_laid_off) as sum_laid_off_per_country
FROM layoffs_staging2
WHERE country = 'United States'
GROUP BY location
ORDER BY 2 DESC
LIMIT 10;

 * mysql+pymysql://root:***@localhost:3306/worldlayoffs
10 rows affected.


[('SF Bay Area', Decimal('125601')),
 ('Seattle', Decimal('34743')),
 ('New York City', Decimal('29364')),
 ('Boston', Decimal('10785')),
 ('Austin', Decimal('8980')),
 ('Chicago', Decimal('6419')),
 ('Los Angeles', Decimal('6415')),
 ('Phoenix', Decimal('5067')),
 ('Philadelphia', Decimal('2989')),
 ('Miami', Decimal('2425'))]

### 5. Company stage with the most layoffs

In [11]:
%%sql
SELECT 
    stage,
    SUM(total_laid_off) as sum_laid_off_per_year
FROM layoffs_staging2
GROUP BY stage
ORDER BY 2 DESC;

 * mysql+pymysql://root:***@localhost:3306/worldlayoffs
17 rows affected.


[('Post-IPO', Decimal('204132')),
 ('Unknown', Decimal('40716')),
 ('Acquired', Decimal('27576')),
 ('Series C', Decimal('20017')),
 ('Series D', Decimal('19225')),
 ('Series B', Decimal('15311')),
 ('Series E', Decimal('12697')),
 ('Series F', Decimal('9932')),
 ('Private Equity', Decimal('7957')),
 ('Series H', Decimal('7244')),
 ('Series A', Decimal('5678')),
 ('Series G', Decimal('3697')),
 ('Series J', Decimal('3570')),
 ('Series I', Decimal('2855')),
 ('Seed', Decimal('1636')),
 ('Subsidiary', Decimal('1094')),
 (None, Decimal('322'))]

The stage column shows the stage at which the company is in. The POST-IPO experienced the highest layoffs with a total of 408,264 layoffs within the 3 years. POST-IPO(Initial Public Offering) are large companies such as Google and Amazon. The least affected stage was the Subsidiary company stage. 

### 6. Top 10 industries with the most layoffs

In [12]:
%%sql
SELECT industry, 
    SUM(total_laid_off) as sum_laid_off_per_industry
FROM layoffs_staging2
GROUP BY industry
ORDER BY 2 DESC
LIMIT 10;

 * mysql+pymysql://root:***@localhost:3306/worldlayoffs
10 rows affected.


[('Consumer', Decimal('45182')),
 ('Retail', Decimal('43613')),
 ('Other', Decimal('36289')),
 ('Transportation', Decimal('33748')),
 ('Finance', Decimal('28344')),
 ('Healthcare', Decimal('25953')),
 ('Food', Decimal('22855')),
 ('Real Estate', Decimal('17565')),
 ('Travel', Decimal('17159')),
 ('Hardware', Decimal('13828'))]

The consumer and retail industries were the most affected by the layoffs with a total of 90,364 and 87,226 layoffs respectively. This could be caused by the covid-19 which led to a lot of shops being closed down during quarantine. 

### 7. Total layoffs per year

In [13]:
%%sql
SELECT 
    YEAR(`date`) as year,
    SUM(total_laid_off) as sum_laid_off_per_year
FROM layoffs_staging2
WHERE YEAR(`date`) IS NOT NULL
GROUP BY YEAR(`date`)
ORDER BY 2 DESC;

 * mysql+pymysql://root:***@localhost:3306/worldlayoffs
4 rows affected.


[(2022, Decimal('160661')),
 (2023, Decimal('125677')),
 (2020, Decimal('80998')),
 (2021, Decimal('15823'))]

The year 2022 experienced the most layoffs, with 160,661 employees being laid off, while 2021 experienced the minimum layoffs.

### 8. Top 5 Companies with maximum layoffs yearwise

In [14]:
%%sql
WITH company_year (company, years, total_laid_off) as
(
    SELECT company,
        YEAR(`date`),
        SUM(total_laid_off)
    FROM layoffs_staging2
    GROUP BY company, YEAR(`date`)
), 
company_year_rank as 
(
    SELECT *,
        DENSE_RANK() OVER(PARTITION BY years ORDER BY total_laid_off DESC) as ranking
    FROM company_year
    WHERE years IS NOT NULL
)
SELECT * 
FROM company_year_rank
WHERE ranking <= 5;

 * mysql+pymysql://root:***@localhost:3306/worldlayoffs
22 rows affected.


[('Uber', 2020, Decimal('7525'), 1),
 ('Booking.com', 2020, Decimal('4375'), 2),
 ('Groupon', 2020, Decimal('2800'), 3),
 ('Swiggy', 2020, Decimal('2250'), 4),
 ('Airbnb', 2020, Decimal('1900'), 5),
 ('Bytedance', 2021, Decimal('3600'), 1),
 ('Katerra', 2021, Decimal('2434'), 2),
 ('Zillow', 2021, Decimal('2000'), 3),
 ('Instacart', 2021, Decimal('1877'), 4),
 ('WhiteHat Jr', 2021, Decimal('1800'), 5),
 ('Meta', 2022, Decimal('11000'), 1),
 ('Amazon', 2022, Decimal('10150'), 2),
 ('Cisco', 2022, Decimal('4100'), 3),
 ('Peloton', 2022, Decimal('4084'), 4),
 ('Carvana', 2022, Decimal('4000'), 5),
 ('Philips', 2022, Decimal('4000'), 5),
 ('Google', 2023, Decimal('12000'), 1),
 ('Microsoft', 2023, Decimal('10000'), 2),
 ('Ericsson', 2023, Decimal('8500'), 3),
 ('Amazon', 2023, Decimal('8000'), 4),
 ('Salesforce', 2023, Decimal('8000'), 4),
 ('Dell', 2023, Decimal('6650'), 5)]

In 2020, Uber experienced the most layoffs with a total of 15,050 layoffs. Bytendance experienced maximum layoffs in 2021, Meta experienced the most layoffs in 2022, while Google experienced the most layoffs in 2023. 

### 9. Monthly layoffs rolling total

In [15]:
%%sql
WITH Rolling_total AS (
    SELECT 
        SUBSTRING(`date`,1,7) AS `Month`,
        SUM(total_laid_off) AS sum_laid_off_per_month
    FROM layoffs_staging2
    WHERE SUBSTRING(`date`,1,7) IS NOT NULL
    GROUP BY `Month`
    ORDER BY 1 ASC
    )
SELECT 
    `Month`,
    sum_laid_off_per_month,
    SUM(sum_laid_off_per_month) OVER(
        ORDER BY `Month`) AS rolling_totals
FROM Rolling_total
LIMIT 20;

 * mysql+pymysql://root:***@localhost:3306/worldlayoffs
20 rows affected.


[('2020-03', Decimal('9628'), Decimal('9628')),
 ('2020-04', Decimal('26710'), Decimal('36338')),
 ('2020-05', Decimal('25804'), Decimal('62142')),
 ('2020-06', Decimal('7627'), Decimal('69769')),
 ('2020-07', Decimal('7112'), Decimal('76881')),
 ('2020-08', Decimal('1969'), Decimal('78850')),
 ('2020-09', Decimal('609'), Decimal('79459')),
 ('2020-10', Decimal('450'), Decimal('79909')),
 ('2020-11', Decimal('237'), Decimal('80146')),
 ('2020-12', Decimal('852'), Decimal('80998')),
 ('2021-01', Decimal('6813'), Decimal('87811')),
 ('2021-02', Decimal('868'), Decimal('88679')),
 ('2021-03', Decimal('47'), Decimal('88726')),
 ('2021-04', Decimal('261'), Decimal('88987')),
 ('2021-06', Decimal('2434'), Decimal('91421')),
 ('2021-07', Decimal('80'), Decimal('91501')),
 ('2021-08', Decimal('1867'), Decimal('93368')),
 ('2021-09', Decimal('161'), Decimal('93529')),
 ('2021-10', Decimal('22'), Decimal('93551')),
 ('2021-11', Decimal('2070'), Decimal('95621'))]